src/ngResource/resource.js

```diff
@@ -800,25 +857,8 @@ angular.module('ngResource', ['ng']).
               }
             });
 
-            promise = promise.then(
-              function(response) {
-                var value = responseInterceptor(response);
-                (success || noop)(value, response.headers, response.status, response.statusText);
-                return value;
-              },
-              (hasError || hasResponseErrorInterceptor) ?
-                function(response) {
-                  if (hasError && !hasResponseErrorInterceptor) {
-                    // Avoid `Possibly Unhandled Rejection` error,
-                    // but still fulfill the returned promise with a rejection
-                    promise.catch(noop);
-                  }
-                  if (hasError) error(response);
-                  return hasResponseErrorInterceptor ?
-                    responseErrorInterceptor(response) :
-                    $q.reject(response);
-                } :
-                undefined);
+            // Run the `success`/`error` callbacks, but do not let them affect the returned promise.
+            promise.then(successCallback, errorCallback);
```

Comment:
So you explicitly choose for the success and error callbacks not to affect the returned promise, thus causing a breaking change.

Is this BC really necessary? Why not allow them to pass through to returned promise, given that this was the previous behaviour?

Reply:
Just to be clear, this commit will be a BC regardless if we propagate callback errors to `$promise` or not.

Previously, callbacks did not 100% affect `$promise`. E.g. their returned value was ignored. It just so happened that throwing was reflected on `$promise` as a rejection (since it happened inside a `then()` callback).

Changing the general behavior (e.g. forwarding the value returned by callbacks to `$promise`) would be unexpected and confusing imo. I am fine with both the current state of the PR as well as having thrown errors propagate to `$promise` as rejections. (I just chose to not preserve that particular aspect of the previous behavior, because it felt unintentional.)

Reply:
+1